### ISSUE TITLE
repe cmpsb issue?

### ISSUE THREAD

**Issue by Owlz** [angr/simuvex#4]
Not sure if this is an issue or if I haven't set it up correctly. I've been trying to use Angr to complete the HITCON NPC challenge. When simulating through, I noticed that paths hitting the "repe cmpsb" at 0x400c9a were spinning MANY times (sometimes upwards of 600 before I stopped it). The actual assembly instruction should have only repeated 7 times.

There is a "rep stosq" in the path before it, and that rep instruction does loop the number of times that it should. This leads me to think that somehow the implementation of repe cmpsb in simuvex isn't accurate.

**Comment by zardus** [comment]
I'd be surprised if it was an IR translation issue, as that would be an
error in VEX. Those do happen, but if I had to guess, I'd say that this is
probably due to some extra symbolic data introduced into the environment
from a poorly modeled syscall or something.

NPC was a very very bad challenge for angr. I doubt (though I'm not sure)
that even translating the checks into z3 by hand would yield good results.

Could you upload a test case that triggers this behavior?

**Comment by Owlz** [comment]
import angr
import simuvex

def strdup(state):
        malloc = simuvex.SimProcedures['libc.so.6']['malloc']
        malloc_out = malloc(state,inline=True,arguments=[state.BVV(1024,32)])
        state.regs.rax = malloc_out.ret_expr

def strlen(state):
        # Force a length of 88
        state.regs.rax = 88

b = angr.Project("npc")

b.hook(0x400BDE,strdup,length=5)
b.hook(0x400BF4,strlen,length=5)

# Start at main
initial_state = b.factory.entry_state(addr=0x400BC0,args=[angr.StringSpec(string="npc"), angr.StringSpec(sym_length=8, nonnull=True)])

avoid=(0x401FB0,0x401E24,0x401E42)

pg = b.factory.path_group(initial_state,immutable=False)

pg.explore(find=(0x400EF5),avoid=avoid)

Basically I just let that run for a while, then ctrl-c and check the backtrace of the active paths.

**Comment by ltfish** [comment]
`npc` is a challenge that angr will never solve :-)
